README.md
# Prometheus Oracle Exporter

# Description

A [Prometheus](https://prometheus.io/) exporter for Oracle. All requests are launched in parallel processes and do not block the HTTP entry point of Prometheus.

# Installation

## Docker

From the very beginning, you need to create a configuration file with a list of all connections and all the requests that you need, an example file in `config.yaml.template`.
Just copy to `config.yaml` and fill it.

You can run via Docker using an existing image. If you don't already have an Oracle server, you can run one locally in a container and then link the exporter to it.

```bash
docker run -d --name oracle -p 1521:1521 wnameless/oracle-xe-11g:16.04
docker run -d -v /Users/username/prometheus_oracle/config.yaml:/config.yaml --name prometheus_oracle --link=oracle -p 9101:9101 juev/prometheus_oracle
```

Or just:

```bash
docker run --rm -v /Users/username/prometheus_oracle/config.yaml:/config.yaml --name prometheus_oracle -p 9101:9101 juev/prometheus_oracle
```

# Metrics

The following metrics are exposed currently.

- oracledb_exporter_dbmetric
- oracledb_exporter_query_duration
- oracledb_exporter_query_duration_overhead
- oracledb_exporter_query_error
- oracledb_exporter_string_dbmetric
- oracledb_exporter_up

Example:

```
# HELP oracledb_exporter_dbmetric Value of Business metrics from Database
# TYPE oracledb_exporter_dbmetric gauge
oracledb_exporter_dbmetric{database="DB1",name="metric1"} 1.59862528e+06
oracledb_exporter_dbmetric{database="DB1",name="metric2"} 0
# HELP oracledb_exporter_query_duration Duration of the query in seconds
# TYPE oracledb_exporter_query_duration gauge
oracledb_exporter_query_duration{database="DB1",name="metric1"} 0.005608735
oracledb_exporter_query_duration{database="DB1",name="metric2"} 0.008431937
oracledb_exporter_query_duration{database="DB2",name="metric3"} 0.031262699
oracledb_exporter_query_duration{database="DB2",name="metric4"} 0.01056187
oracledb_exporter_query_duration{database="DB2",name="metric5"} 0.006411418
oracledb_exporter_query_duration{database="DB2",name="metric6"} 0.00876666
oracledb_exporter_query_duration{database="DB2",name="metric7"} 0.027751112
oracledb_exporter_query_duration{database="DB2",name="metric8"} 0.008800151
oracledb_exporter_query_duration{database="DB2",name="metric9"} 0.012350262
# HELP oracledb_exporter_query_duration_overhead Overhead in the duration
# TYPE oracledb_exporter_query_duration_overhead gauge
oracledb_exporter_query_duration_overhead{database="DB1",name="metric1"} 0
oracledb_exporter_query_duration_overhead{database="DB1",name="metric2"} 0
oracledb_exporter_query_duration_overhead{database="DB2",name="metric3"} 0
oracledb_exporter_query_duration_overhead{database="DB2",name="metric4"} 0
oracledb_exporter_query_duration_overhead{database="DB2",name="metric5"} 0
oracledb_exporter_query_duration_overhead{database="DB2",name="metric6"} 0
oracledb_exporter_query_duration_overhead{database="DB2",name="metric7"} 0
oracledb_exporter_query_duration_overhead{database="DB2",name="metric8"} 0
oracledb_exporter_query_duration_overhead{database="DB2",name="metric9"} 0
# HELP oracledb_exporter_query_error Result of last query, 1 if we have errors on running query
# TYPE oracledb_exporter_query_error gauge
oracledb_exporter_query_error{database="DB1",name="metric1"} 0
oracledb_exporter_query_error{database="DB1",name="metric2"} 0
oracledb_exporter_query_error{database="DB2",name="metric3"} 0
oracledb_exporter_query_error{database="DB2",name="metric4"} 0
oracledb_exporter_query_error{database="DB2",name="metric5"} 0
oracledb_exporter_query_error{database="DB2",name="metric6"} 0
oracledb_exporter_query_error{database="DB2",name="metric7"} 0
oracledb_exporter_query_error{database="DB2",name="metric8"} 0
oracledb_exporter_query_error{database="DB2",name="metric9"} 0
# HELP oracledb_exporter_string_dbmetric Value of Business metrics from Database, using string value
# TYPE oracledb_exporter_string_dbmetric gauge
oracledb_exporter_string_dbmetric{database="DB2",name="metric3",value="0"} 0
oracledb_exporter_string_dbmetric{database="DB2",name="metric4",value="0"} 0
oracledb_exporter_string_dbmetric{database="DB2",name="metric5",value="0"} 0
oracledb_exporter_string_dbmetric{database="DB2",name="metric6",value="1 paid errors"} 1
oracledb_exporter_string_dbmetric{database="DB2",name="metric7",value="Error in logs"} 1
oracledb_exporter_string_dbmetric{database="DB2",name="metric8",value="Please call me"} 1
oracledb_exporter_string_dbmetric{database="DB2",name="metric9",value="0"} 0
# HELP oracledb_exporter_up Database status
# TYPE oracledb_exporter_up gauge
oracledb_exporter_up{database="DB1"} 1
oracledb_exporter_up{database="DB2"} 1
```

# Config file

```
# Host, default value `0.0.0.0` (optional)
host: 0.0.0.0
# Port, default value `9101` (optional)
port: 9101
# QueryTimeout, default value `30` in seconds (optional)
queryTimeout: 5

# Array of databases and queries
databases:
    # Host, default value `127.0.0.1`, hostname for DB (optional)
  - host: 'dummy'
    # User (required)
    user: dummy
    # Port, default value `1522` (optional)
    port: 1522
    # Password (required)
    password: 'password'
    # Database name (required)
    database: dummy
    # MaxIdleConns, default value `10` (optional)
    maxIdleConns: 10
    # MaxOpenConns, default value `10` (optional)
    maxOpenConns: 10
    # Aray of queries
    queries:
        # SQL, query (required)
      - sql: "select numbers1 from dummy"
        # SQL name (required)
        name: value1
        # Interval between queries, default value `1` in minites (optional)
        interval: 1
        # Type of query, default value `value` (can be value/string) (optional)
        type: "value"
```

CHANGELOG.md
# Changelog


<a name="v.0.3.0"></a>
## [v.0.3.0](https://github.com/juev/prometheus_oracle/compare/v0.2.0...v.0.3.0) (2020-02-20)


<a name="v0.2.0"></a>
## [v0.2.0](https://github.com/juev/prometheus_oracle/compare/0.1.0...v0.2.0) (2020-02-19)


<a name="0.1.0"></a>
## 0.1.0 (2020-02-12)



main.go
package main

import (
	"context"
	"database/sql"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/jasonlvhit/gocron"
	"github.com/kkyr/fig"
	"github.com/mattn/go-oci8"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/sirupsen/logrus"
	flag "github.com/spf13/pflag"
)

type Configuration struct {
	Host         string `fig:"host,default=0.0.0.0"`
	Port         string `fig:"port,default=9101"`
	QueryTimeout string `fig:"querytimeout,default=30"`
	Databases    []Database
}

type Database struct {
	Dsn          string
	Host         string  `fig:",default=127.0.0.1"`
	User         string  `fig:"user"`
	Password     string  `fig:"password"`
	Database     string  `fig:"database"`
	Port         string  `fig:"port,default=1522"`
	MaxIdleConns string  `fig:",default=10"`
	MaxOpenConns string  `fig:",default=10"`
	Queries      []Query `fig:"queries"`
	db           *sql.DB
}

type Query struct {
	Sql      string `fig:"sql"`
	Name     string `fig:"name"`
	Interval string `fig:"interval,default=1"`
	Type     string `fig:"type,default=value"`
}

const (
	namespace = "oracledb"
	exporter  = "exporter"
)

var (
	configuration Configuration
	metricMap     map[string]*prometheus.GaugeVec
	timeout       int
	maxIdleConns  int
	maxOpenConns  int
	err           error
	configFile    string
	logFile       string
)

func init() {
	flag.StringVarP(&configFile, "configFile", "c", "config.yaml", "Config file name (default: config.yaml)")
	flag.StringVarP(&logFile, "logFile", "l", "stdout", "Log filename (default: stdout)")

	metricMap = map[string]*prometheus.GaugeVec{
		"value": prometheus.NewGaugeVec(prometheus.GaugeOpts{
			Namespace: namespace,
			Subsystem: exporter,
			Name:      "dbmetric",
			Help:      "Value of Business metrics from Database",
		}, []string{"database", "name"}),
		"string": prometheus.NewGaugeVec(prometheus.GaugeOpts{
			Namespace: namespace,
			Subsystem: exporter,
			Name:      "string_dbmetric",
			Help:      "Value of Business metrics from Database, using string value",
		}, []string{"database", "name", "value"}),
		"result": prometheus.NewGaugeVec(prometheus.GaugeOpts{
			Namespace: namespace,
			Subsystem: exporter,
			Name:      "query_error",
			Help:      "Result of last query, 1 if we have errors on running query",
		}, []string{"database", "name"}),
		"duration": prometheus.NewGaugeVec(prometheus.GaugeOpts{
			Namespace: namespace,
			Subsystem: exporter,
			Name:      "query_duration_seconds",
			Help:      "Duration of the query in seconds",
		}, []string{"database", "name"}),
		"overhead": prometheus.NewGaugeVec(prometheus.GaugeOpts{
			Namespace: namespace,
			Subsystem: exporter,
			Name:      "query_duration_overhead",
			Help:      "Overhead in the duration",
		}, []string{"database", "name"}),
		"up": prometheus.NewGaugeVec(prometheus.GaugeOpts{
			Namespace: namespace,
			Subsystem: exporter,
			Name:      "up",
			Help:      "Database status",
		}, []string{"database"}),
	}
	for _, metric := range metricMap {
		prometheus.MustRegister(metric)
	}
}

func execQuery(database Database, query Query) {

	defer func(begun time.Time) {
		duration := time.Since(begun).Seconds()
		metricMap["duration"].WithLabelValues(database.Database, query.Name).Set(duration)
		interval, _ := strconv.Atoi(query.Interval)
		// duration in secons, interval in minutes
		if duration > float64(interval*60) {
			metricMap["overhead"].WithLabelValues(database.Database, query.Name).Set(1)
		} else {
			metricMap["overhead"].WithLabelValues(database.Database, query.Name).Set(0)
		}
	}(time.Now())

	// Reconnect if we lost connection
	if err := database.db.Ping(); err != nil {
		if strings.Contains(err.Error(), "sql: database is closed") {
			logrus.Infoln("Reconnecting to DB: ", database.Database)
			database.db, _ = sql.Open("oci8", database.Dsn)
			database.db.SetMaxIdleConns(maxIdleConns)
			database.db.SetMaxOpenConns(maxOpenConns)
		}
	}

	// Validate connection
	if err := database.db.Ping(); err != nil {
		logrus.Errorf("Error on connect to database '%s': %v", database.Database, err)
		metricMap["up"].WithLabelValues(database.Database).Set(0)
		return
	} else {
		metricMap["up"].WithLabelValues(database.Database).Set(1)
	}

	// query db
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
	defer cancel()
	rows, err := database.db.QueryContext(ctx, query.Sql)
	if ctx.Err() == context.DeadlineExceeded {
		logrus.Errorf("oracle query '%s' timed out", query.Name)
		metricMap["result"].WithLabelValues(database.Database, query.Name).Set(1)
		return
	}
	if err != nil {
		logrus.Errorf("oracle query '%s' failed: %v", query.Name, err)
		metricMap["result"].WithLabelValues(database.Database, query.Name).Set(1)
		return
	}

	cols, _ := rows.Columns()
	vals := make([]interface{}, len(cols))
	defer func() {
		err := rows.Close()
		if err != nil {
			logrus.Fatal(err)
		}
	}()

	for rows.Next() {
		for i := range cols {
			vals[i] = &vals[i]
		}

		err = rows.Scan(vals...)
		if err != nil {
			break
		}

		for i := range cols {
			if vals[i] == nil {
				if query.Type == "string" {
					metricMap["string"].WithLabelValues(database.Database, query.Name, "0").Set(0)
				} else {
					metricMap["value"].WithLabelValues(database.Database, query.Name).Set(0)
				}
			} else {
				if query.Type == "string" {
					metricMap["string"].WithLabelValues(database.Database, query.Name, vals[i].(string)).Set(1)
				} else {
					val, err := strconv.ParseFloat(strings.TrimSpace(vals[i].(string)), 64)
					if err != nil {
						logrus.Errorf("Cannot convert value '%s' to float on query '%s': %v", vals[i].(string), query.Name, err)
						metricMap["result"].WithLabelValues(database.Database, query.Name).Set(1)
						return
					}
					metricMap["value"].WithLabelValues(database.Database, query.Name).Set(val)
				}
			}
		}
		metricMap["result"].WithLabelValues(database.Database, query.Name).Set(0)
	}
}

func main() {
	flag.Parse()
	if logFile == "stdout" {
		logrus.SetOutput(os.Stdout)
	} else {
		file, err := os.OpenFile(logFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
		if err == nil {
			logrus.SetOutput(file)
		} else {
			logrus.Info("Failed to log to file, using default stdout")
			logrus.SetOutput(os.Stdout)
		}
	}

	err = fig.Load(&configuration, fig.File(configFile))
	if err != nil {
		logrus.Fatal("Fatal error on reading configuration: ", err)
	}

	//indent, err := json.MarshalIndent(configuration, "", "    ")
	//if err != nil {
	//	logrus.Error(err)
	//}
	//fmt.Println(string(indent))

	timeout, err = strconv.Atoi(configuration.QueryTimeout)
	if err != nil {
		logrus.Fatal("error while converting timeout option value: ", err)
		//panic(err)
	}

	for _, database := range configuration.Databases {
		// connect to database
		database.Dsn = oci8.QueryEscape(database.User) + "/" + oci8.QueryEscape(database.Password) +
			"@" + database.Host + ":" + database.Port + "/" + database.Database
		logrus.Infoln("Connecting to DB: ", database.Database)
		database.db, err = sql.Open("oci8", database.Dsn)
		if err != nil {
			logrus.Errorln("Error connecting to db: ", err)
		}
		maxIdleConns, err = strconv.Atoi(database.MaxIdleConns)
		if err != nil {
			logrus.Fatal("error while converting maxIdleConns option value: ", err)
			//panic(err)
		}

		maxOpenConns, err = strconv.Atoi(database.MaxOpenConns)
		if err != nil {
			logrus.Fatal("error while converting maxOpenConns option value: ", err)
			//panic(err)
		}

		database.db.SetMaxIdleConns(maxOpenConns)
		database.db.SetMaxOpenConns(maxOpenConns)

		// create cron jobs for every query on database
		if err := database.db.Ping(); err == nil {
			for _, query := range database.Queries {
				gocron.Every(5).Minutes().DoSafely(execQuery, database, query)
			}
		} else {
			logrus.Errorf("Error connecting to db '%s': %v", database.Database, err)
		}
	}

	gocron.Start()
	gocron.RunAll()

	prometheusConnection := configuration.Host + ":" + configuration.Port
	logrus.Printf("listen: %s", prometheusConnection)
	http.Handle("/metrics", promhttp.Handler())
	err = http.ListenAndServe(prometheusConnection, nil)
	if err != nil {
		logrus.Fatalln("Fatal error on serving metrics:", err)
	}
}
